File: services/ml/src/db/models/source_records.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Uuid
from sqlalchemy.orm import Mapped, mapped_column

from src.db.base import Base


class StudentFinancialRecord(Base):
    __tablename__ = "student_financial_records"

    id: Mapped[uuid.UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True)
    student_profile_id: Mapped[uuid.UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("student_profiles.id", ondelete="CASCADE"),
        nullable=False,
    )
    income_in_rs: Mapped[float] = mapped_column(Float, nullable=False)
    land_owned_acres: Mapped[float] = mapped_column(Float, nullable=False)
    vehicles_owned: Mapped[int] = mapped_column(Integer, nullable=False)
    electricity_consumption: Mapped[float] = mapped_column(Float, nullable=False)
    pending_loans: Mapped[int] = mapped_column(Integer, nullable=False)
    business_ownership: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )


class StudentSocialRecord(Base):
    __tablename__ = "student_social_records"

    id: Mapped[uuid.UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True)
    student_profile_id: Mapped[uuid.UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("student_profiles.id", ondelete="CASCADE"),
        nullable=False,
    )
    caste: Mapped[str] = mapped_column(String, nullable=False)
    father_caste: Mapped[str] = mapped_column(String, nullable=False)
    avg_caste_population_per: Mapped[float] = mapped_column(Float, nullable=False)
    officer_approvals_per_day: Mapped[float] = mapped_column(Float, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )


class StudentTransactionSummary(Base):
    __tablename__ = "student_transaction_summaries"

    id: Mapped[uuid.UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True)
    student_profile_id: Mapped[uuid.UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("student_profiles.id", ondelete="CASCADE"),
        nullable=False,
    )
    weekly_spending: Mapped[float] = mapped_column(Float, nullable=False)
    monthly_spending: Mapped[float] = mapped_column(Float, nullable=False)
    transaction_count: Mapped[int] = mapped_column(Integer, nullable=False)
    avg_transaction_value: Mapped[float] = mapped_column(Float, nullable=False)
    luxury_items_bought: Mapped[int] = mapped_column(Integer, nullable=False)
    weekend_spending_ratio: Mapped[float] = mapped_column(Float, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )


class StudentMedicalSummary(Base):
    __tablename__ = "student_medical_summaries"

    id: Mapped[uuid.UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True)
    student_profile_id: Mapped[uuid.UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("student_profiles.id", ondelete="CASCADE"),
        nullable=False,
    )
    hospital_visits_per_year: Mapped[int] = mapped_column(Integer, nullable=False)
    claim_frequency: Mapped[int] = mapped_column(Integer, nullable=False)
    medical_claim_amount: Mapped[float] = mapped_column(Float, nullable=False)
    avg_claim_amount: Mapped[float] = mapped_column(Float, nullable=False)
    chronic_disease: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )

File: packages/db/schema/source-records.ts
import {
  index,
  integer,
  pgTable,
  real,
  text,
  timestamp,
  uuid,
} from "drizzle-orm/pg-core"
import { studentProfiles } from "./student-profiles"

export const studentFinancialRecords = pgTable(
  "student_financial_records",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    studentProfileId: uuid("student_profile_id")
      .notNull()
      .references(() => studentProfiles.id, { onDelete: "cascade" }),
    incomeInRs: real("income_in_rs").notNull(),
    landOwnedAcres: real("land_owned_acres").notNull(),
    vehiclesOwned: integer("vehicles_owned").notNull(),
    electricityConsumption: real("electricity_consumption").notNull(),
    pendingLoans: integer("pending_loans").notNull(),
    businessOwnership: integer("business_ownership").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("student_financial_records_profile_created_at_idx").on(
      table.studentProfileId,
      table.createdAt,
    ),
  ],
)

export const studentSocialRecords = pgTable(
  "student_social_records",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    studentProfileId: uuid("student_profile_id")
      .notNull()
      .references(() => studentProfiles.id, { onDelete: "cascade" }),
    caste: text("caste").notNull(),
    fatherCaste: text("father_caste").notNull(),
    avgCastePopulationPer: real("avg_caste_population_per").notNull(),
    officerApprovalsPerDay: real("officer_approvals_per_day").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("student_social_records_profile_created_at_idx").on(
      table.studentProfileId,
      table.createdAt,
    ),
  ],
)

export const studentTransactionSummaries = pgTable(
  "student_transaction_summaries",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    studentProfileId: uuid("student_profile_id")
      .notNull()
      .references(() => studentProfiles.id, { onDelete: "cascade" }),
    weeklySpending: real("weekly_spending").notNull(),
    monthlySpending: real("monthly_spending").notNull(),
    transactionCount: integer("transaction_count").notNull(),
    avgTransactionValue: real("avg_transaction_value").notNull(),
    luxuryItemsBought: integer("luxury_items_bought").notNull(),
    weekendSpendingRatio: real("weekend_spending_ratio").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("student_transaction_summaries_profile_created_at_idx").on(
      table.studentProfileId,
      table.createdAt,
    ),
  ],
)

export const studentMedicalSummaries = pgTable(
  "student_medical_summaries",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    studentProfileId: uuid("student_profile_id")
      .notNull()
      .references(() => studentProfiles.id, { onDelete: "cascade" }),
    hospitalVisitsPerYear: integer("hospital_visits_per_year").notNull(),
    claimFrequency: integer("claim_frequency").notNull(),
    medicalClaimAmount: real("medical_claim_amount").notNull(),
    avgClaimAmount: real("avg_claim_amount").notNull(),
    chronicDisease: integer("chronic_disease").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("student_medical_summaries_profile_created_at_idx").on(
      table.studentProfileId,
      table.createdAt,
    ),
  ],
)
